### DISCUSSION TITLE
Just process replacement

### DISCUSSION THREAD

**Discussion by dmringo** [casey/just#1140] [Ideas]
I use `just` occasionally to launch long-running (days/weeks+) processes. I also keep `just` updated pretty regularly, but run into issues if one of these long running processes is alive (can't overwrite the binary during update). A nice feature to work around this kind of thing would be some kind of syntax that would cause `just` to `exec` (in the syscall sense) the appropriate interpreter.

I'm not entirely sure what process creation in Rust looks like for Windows, so it may just not be viable from the start for that reason. I think it would also impose limitations on dependencies (can't really `exec` a recipe dependency and then return to run the dependent recipe) so maybe it would only make sense as a command line option that applies to the root recipe(s) in a given invocation of `just`.

I'd love to hear anyone's thoughts on this: value, feasibility, whatever. :-)

**Comment by casey** [comment]
One possible solution is to move the `just` binary out of the way before upgraded it. You can't overwrite a running binary, but you can move it.

**Reply by dmringo** [reply]
That's fair, although I can't make *that move/copy* a part of the long running recipe to account for my lack of foresight :-P

I guess there are plenty of workarounds, but I thought there may be other benefits to have this behavior. The existence of shell `exec` builtins seems to imply as much (no need to keep a process and its resources around if it's just going to exit when the child it launched exits).

**Reply by casey** [reply]
I wonder what the best way to support this would be. Probably some annotation on a recipe that the command should be exec-ed, or an annotation on a line that that particular line should be exec-ed. A line annotation would be more precise, although the syntax would probably be awkward.

**Reply by dmringo** [reply]
If it's per-line, wouldn't that introduce some footguns? For example (Forgive my unimaginative syntax):

```make
foo: bar
  echo "Done foo!"

bar:
  echo "Running bar!"
  !!{{EXECME}} sleep $LONGTIME
  echo "Done bar!"
```

The last line of `bar` would always be unreachable, and trying to `just foo` would only get you the first line of `bar` plus a long sleep.

The more I think about it, it only really makes sense to me as an annotation on a recipe with a shebang (or a single-line recipe, I guess).  Given that, maybe this is an even more limited use case than I thought. I use shebangs a lot in my justfiles, but I have no idea how common that is.

**Reply by casey** [reply]
Yeah, that would be weird. I think I agree that making it a recipe annotation, and only allowing it on a recipe with a single line or a shebang would be best.
